chain/network/src/peer_manager/connected_peers.rs

```diff
@@ -1,50 +1,121 @@
-    /// Time where the connection was established.
-    pub connection_established_time: time::Instant,
+}
+
+/// Contains information relevant to a connected peer.
+pub(crate) struct ConnectedPeer {
+    // TODO(gprusak): addr should be internal, so that ConnectedPeer will become an API of the
+    // PeerActor.
+    pub addr: actix::Addr<PeerActor>,
+
+    pub peer_info: PeerInfo,
+    pub partial_edge_info: PartialEdgeInfo,
+    pub initial_chain_info: PeerChainInfoV2,
+    pub chain_height: AtomicU64,
+
     /// Who started connection. Inbound (other) or Outbound (us).
     pub peer_type: PeerType,
+    /// Time where the connection was established.
+    pub connection_established_time: time::Instant,
+
+    /// Last time requested peers.
+    pub last_time_peer_requested: ArcSwap<time::Instant>,
+    /// Last time we received a message from this peer.
+    pub last_time_received_message: ArcSwap<time::Instant>,
+    /// Connection stats
+    pub stats: ArcSwap<Stats>,
+
     /// A helper data structure for limiting reading, reporting stats.
     pub throttle_controller: ThrottleController,
-    /// Encoding used for communication.
-    pub encoding: Option<Encoding>,
     pub send_accounts_data_demux: demux::Demux<Vec<Arc<SignedAccountData>>, ()>,
 }
 
+impl fmt::Debug for ConnectedPeer {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        f.debug_struct("ConnectedPeer")
+            .field("peer_info", &self.peer_info)
+            .field("partial_edge_info", &self.partial_edge_info)
+            .field("peer_type", &self.peer_type)
+            .field("connection_established_time", &self.connection_established_time)
+            .finish()
+    }
+}
+
 impl ConnectedPeer {
+    pub fn full_peer_info(&self) -> FullPeerInfo {
+        let mut chain_info = self.initial_chain_info.clone();
+        chain_info.height = self.chain_height.load(Ordering::Relaxed);
+        FullPeerInfo {
+            peer_info: self.peer_info.clone(),
+            chain_info,
+            partial_edge_info: self.partial_edge_info.clone(),
+        }
+    }
+
+    pub fn ban(&self, ban_reason: ReasonForBan) {
+        self.addr.do_send(PeerManagerRequestWithContext {
+            msg: PeerManagerRequest::BanPeer(ban_reason),
+            context: Span::current().context(),
+        });
+    }
+
+    pub fn unregister(&self) {
+        self.addr.do_send(PeerManagerRequestWithContext {
+            msg: PeerManagerRequest::UnregisterPeer,
+            context: Span::current().context(),
+        });
+    }
+
+    pub fn send_message(&self, msg: Arc<PeerMessage>) {
+        let msg_kind = msg.msg_variant().to_string();
+        tracing::trace!(target: "network", ?msg_kind, "Send message");
+        // Sending may fail in case a peer connection is closed in the meantime.
+        if let Err(err) =
+            self.addr.try_send(SendMessage { message: msg, context: Span::current().context() })
```

Comment:
unrelated to the pr, but I wonder if we should just `do_send` here? `try_send` returns `Mailbox::Full` error under some circumstances, but I wasn't able to tell when exactly that happens.

Reply:
Wait, no, this is not unrelated, this `try_send` is a new behavior, right? Before, we were doing either `do_send`, or `send().await`. 


I think this is incorrect: `try_send` is a non-blocking send, it can return with `WouldBlock`. So this essentially means we can drop messages at the actor layer, which feels wrong. 

I'd say just `do_send` here: I think we care about not droping the messages, but dont' care as much about within-actix automated flow control

Reply:
it was a send() which could fail only if the queue was full. So I replaced it with try_send so that we don't have to await the send, while keeping the semantics of failing if mailbox is full or closed.

Reply:
do_send is fine with me as well

Reply:
done

Reply:
>it was a send() which could fail only if the queue was full. 

`send()` does not fail if the queue is full, it blocks. It only fails if the queue is closed or if the caller used `send_timeout`

Reply:
The error type is different: MailboxError vs SendError

Reply:
true, I missed that. Ok, so the difference between send and do_send will be that the message will be silently dropped in case the mailbox is closed.

Reply:
Hm, it seems we are still not at the same page. In my mind, the main difference is that we drop a messsage if the mailbox is full in one case, and just wait and eventually deliver the message in another. 

When the mailbox is closed (= actor is dead) the behavior is the same -- no message delivered.

Reply:
try_send, inbox full/closed => error
send, inbox closed => error
do_send, no error

in case of send() we were able to log the fact that message was dropped when actor is dead.
with do_send we cannot do that.

Reply:
so, should I go back to send()?

Reply:
it depends, there might me a race condition when a peer disconnects and we try to send a message to it.

Reply:
I think this is going to be just business as usual -- in general, peers can come and go at any moment, closed mailbox in this case is not a programming error.

Reply:
ok, staying with do_send